public/functions/geoid.js

```diff
@@ -0,0 +1,292 @@
+        var v00 = null;
+        var v01 = null;
+        var v10 = null;
+        var v11 = null;
+        var t = null;
+
+        return function (lat, lon, cubic) {
+            if (lon < 0) lon += 360;
+            var fy = (90 - lat) * model.rlatres;
+            var fx = lon * model.rlonres;
+            var iy = Math.floor(fy);
+            var ix = Math.floor(fx);
+            fx -= ix;
+            fy -= iy;
+            if (iy === (model.height - 1)) {
+                iy--;
+            }
+
+            if ((cached_ix !== ix) || (cached_iy !== iy)) {
+                cached_ix = ix;
+                cached_iy = iy;
+                if (cubic) {
+                    var c3x = c3;
+                    var c0x = c0;
+                    if (iy === 0) {
+                        c3x = c3n;
+                        c0x = c0n;
+                    } else if (iy === (model.height - 2)) {
+                        c3x = c3s;
+                        c0x = c0s;
+                    }
+                    var v = [
+                        model.rawval(ix, iy - 1),
+                        model.rawval(ix + 1, iy - 1),
+                        model.rawval(ix - 1, iy),
+                        model.rawval(ix, iy),
+                        model.rawval(ix + 1, iy),
+                        model.rawval(ix + 2, iy),
+                        model.rawval(ix - 1, iy + 1),
+                        model.rawval(ix, iy + 1),
+                        model.rawval(ix + 1, iy + 1),
+                        model.rawval(ix + 2, iy + 1),
+                        model.rawval(ix, iy + 2),
+                        model.rawval(ix + 1, iy + 2)
+                    ];
+                    t = Array.apply(null, Array(10)).map(function (_, i, arr) {
+                        return v.reduce(function (acc, vj, j, arr) {
+                            return acc + vj * c3x[j][i];
+                        }, 0) / c0x;
+                    });
+                } else {
+                    v00 = model.rawval(ix, iy);
+                    v01 = model.rawval(ix + 1, iy);
+                    v10 = model.rawval(ix, iy + 1);
+                    v11 = model.rawval(ix + 1, iy + 1);
+                }
+            }
+
+            var h = null;
+            if (cubic) {
+                h = t[0] +
+                    fx * (t[1] + fx * (t[3] + fx * t[6])) +
+                    fy * (
+                        t[2] + fx * (t[4] + fx * t[7]) +
+                        fy * (t[5] + fx * t[8] + fy * t[9])
+                    );
+            } else {
+                var a = (1 - fx) * v00 + fx * v01;
+                var b = (1 - fx) * v10 + fx * v11;
+                h = (1 - fy) * a + fy * b;
+            }
+
+            return model.offset + model.scale * h;
+        };
+    });
+};
+
+// function that takes in lat, long, and returns the height
+async function getGeoidHeightFromPGM(geoidFilePath, lat, long) {
+    const geoidModel = await loadGeoidModel(geoidFilePath);
```

Comment:
are we loading this every invocation of this function?

Reply:
yes, however it's a "local file" in the google cloud function so it's not downloading it, it is in the same local directory. i honestly don't know how to keep it in memory and reference it, instead of reloading it upon each invocation, and this seems to work in <1 second each time, so I'll just cross my fingers and praise gcp cloud functions that it works